### DISCUSSION TITLE
A more comfy unit test syntax!

### DISCUSSION THREAD

**Discussion by ghost** [vlang/v#10276] [Ideas]
Hello fellow coders!
I've recently been using V _alot_, and I honestly love it! May sound opinionated, but V has the most thought-out syntax, type checker, and OOP-style paradigms between most mainstream programming languages, and that makes it stand out in the most charming ways. Any sane person would think it's 'bout time I'd consider giving back!

Okay, to the point we go.
A while ago, I came across an indie language called [Ziglang](https://ziglang.org). The community seems to have different goals for their language than V, but there are some stuff that I just crave about the syntax!
One nice innovation is their __unit test syntax__.

If you'd look through [their documentation](https://ziglang.org/documentation/master/), its absolutely crammed with statements like `test "test_name" { ... }`. Here's [their elaboration on the test syntax itself](https://ziglang.org/documentation/master/#Zig-Test). Feel free to disagree, but I think this syntax atleast has to be considered to be included.

Here's a basic imagination of how this would work in V:

__Code__ (marked as ziglang code for syntax highlighting)
```zig
module math

const (
    minimum_error = .001
)

test "math.sqrt precision test" {
    test_array := [0.0, 1.9, 1.623, 1931.5024, math.max_f32]

    for i in test_array {
        root := math.sqrt (i)
        restored := root * root
        error := math.abs (restored - root)
        assert error <= minimum_error
    }
}

test "bad test" {
    assert 0 == 1
}
```

__Command line interface__
```bash
$ v test .
Test [1/1] math.sqrt precision test... Success
Test [2/1] bad test... Fail

**stderr here**

1 passed; 1 failed.
```

**Comment by crthpl** [comment]
Note that this will not require a new keyword - we could just detect a `.name` in top level statements, and check if it the literal is `test`. Syntax highlighters should still mark it as a a keyword.

**Comment by medvednikov** [comment]
Right now we use 

```
fn test_sqrt_precision() {
vs
test "math.sqrt precision test" {
```

I think test functions are fine, the description with spaces is of course nicer. @spytheman what do you think?

**Reply by ghost** [reply]
Maybe its a little too weird to think that, but I think it's actually possible and not unacceptable for a programmer to mistakenly create a test function? Some example identifiers I can currently think of are `test_fitness` `test_voltage` `test_user_input` & etc...

Other than nice spaces, I think this style can also be considered more comfortable to read if one were to put tests in the same file as the sources? This way the coder gets to
* look at a function's body whilst writing down tests for it,
* reduce the number of tab switches to the test source, and
* not mistake actual functions with tests bc of different syntax highlighting (already avoided by using different files for tests but hey)

Sorry if this sounds opinionated and rushed, my bad :D
Pardon me if i am not taking some really obvious thing into account, im really fatigué rn 🙃
